pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+            if fixed_covar:
+                fixed_covariates.append(fixed_covar)
+
+        if fixed_covariates:
+            F = np.unique(fixed_covariates)[0]
+            F = "0 + " + F
+            F = np.asarray(patsy.dmatrix(F, df))
+        else:
+            F = None
+
+        X = np.stack(alt_covariates, axis=1).T
+        if X.shape != (n_obs, n_alts, n_covariates):
+            raise ValueError(
+                f"X has shape {X.shape}, but expected {(n_obs, n_alts, n_covariates)}."
+            )
+        return X, F, alts, np.unique(fixed_covariates)
+
+    @staticmethod
+    def _prepare_y_outcome(df, alternatives, depvar):
+        """Encode the outcome category variable for use in the modelling.
+
+        The order of the alterntives should map to the order of the
+        utility formulas.
+        """
+        prod_mapping = dict(zip(alternatives, range(len(alternatives)), strict=False))
+        df["mode_encoded"] = df[depvar].map(prod_mapping)
+        y = df["mode_encoded"].values
+        return y, prod_mapping
+
+    @staticmethod
+    def _parse_nesting(nest_dict, product_indices):
+        """Encode the relevant nesting structure specified when initialising the model.
+
+        We build two dictionaries for the top and middle logit
+        models in the hierarchical nesting structure as required.
+        """
+        if not nest_dict:
+            raise ValueError("Nesting structure must not be empty.")
+        top_level = {}
+        mid_level = {}
+        for k in nest_dict.keys():
+            if isinstance(nest_dict[k], dict):
+                collected_idxs = []
+                for j in nest_dict[k]:
+                    if isinstance(nest_dict[k][j], dict):
+                        raise ValueError("Cannot have more than 2 layers of Nesting")
+                    else:
+                        idxs = [product_indices[i] for i in nest_dict[k][j]]
+                        mid_level[k + "_" + j] = idxs
+                        collected_idxs = collected_idxs + idxs
+                top_level[k] = np.sort(collected_idxs)
+            else:
+                top_level[k] = np.sort([product_indices[i] for i in nest_dict[k]])
+
+        if not mid_level:
+            mid_level = None
+        return top_level, mid_level
+
+    @staticmethod
+    def _prepare_coords(df, alternatives, covariates, f_covariates, nests):
+        """Prepare coordinates for PyMC nested logit model."""
+        if isinstance(f_covariates, np.ndarray) & (f_covariates.size > 0):
+            f_cov = [s.strip() for s in f_covariates[0].split("+")]
+        else:
+            f_cov = []
+        coords = {
+            "alts": alternatives,
+            "alts_probs": alternatives[:-1],
+            "alt_covariates": covariates,
+            "fixed_covariates": f_cov,
+            "nests": nests,
+            "obs": range(len(df)),
+        }
+        return coords
+
+    def preprocess_model_data(
+        self, choice_df, utility_equations
```

Comment:
type hints missing

Reply:
alsmo missing parameter description in docstrings

Reply:
added